### ISSUE TITLE
Clarion Professional Developer 2.1 - Clarion Compiler Issue Migrating from vDosPlus to DOSBox-X

### ISSUE THREAD

**Issue by zydanielson** [joncampbell123/dosbox-x#1892]
@Wengier encouraged me to migrate from vDosPlus to DOSBox-X as the former is no longer being developed by him. So I'm trying to get the Clarion Professional Developer 2.1 to work in DOSBox-X. The community here seems active and reliable. I appreciate that.

There is a problem when running the Clarion Compiler (ccmp.exe) which compiles Clarion source files (.cla) to Clarion processor files (.pro). The resulting .pro file can be run in the Processer module (cpro.exe) for debugging or the Runtime (crun.exe) without the debug overhead.

**The Problem**
The Clarion Compiler (ccmp.exe) runs fine in vDosPlus, but produces errors in DOXBox-X.

Start of compile in vDosPlus:
![cla21compiler_vDosPlus_begin](https://user-images.githubusercontent.com/11687580/94626558-8c69b100-0278-11eb-9933-cb224dcedbd9.jpg)

End of compile in vDosPlus. Compile successful:
![cla21compiler_vDosPlus_end](https://user-images.githubusercontent.com/11687580/94626653-c3d85d80-0278-11eb-8ea2-6b3b94cd82d3.jpg)

Start of compile in DOSBox-X:
![cla21compiler_DOSBox-X_begin](https://user-images.githubusercontent.com/11687580/94626897-4eb95800-0279-11eb-9386-5b0d953a6f8c.jpg)

End of compile in DOSBox-X. Compiler has error messages:
![cla21compiler_DOSBox-X_end1](https://user-images.githubusercontent.com/11687580/94626944-6a246300-0279-11eb-8553-45fb5dcdbe6e.jpg)

DOSBox-X. The first 7 error messages listed. Compile unsuccessful:
![cla21compiler_DOSBox-X_end2](https://user-images.githubusercontent.com/11687580/94627094-d7d08f00-0279-11eb-98a1-75f4a18f3661.jpg)

**Source Files Are from the Same Directory**
I have set up both vDosPlus and DOSBox-X to read and write all Clarion files to/from the same directory, as seen in the start of the compiler screenshots above:  **C:\BKM\Master.cla**. Thus, when I get the bad compile in DOSBox-X, I run it again in vDosPlus, without changing anything, and get a good compile.

**The Time it Takes to compile is different in both Boxes**
- vDosPlus: 3.02 seconds
- DOSBox-X: 1 minute 26.69 seconds

**Where the Greatest Delays are Noticed**
The bottom line of the Clarion Compiler screen reports what it is doing during the compile. One thing is compiling code. Another is when it is processing an include file (.inc) Include files are stored in the same directory as the source file. The greatest delays in time are when the bottom line says it is processing an include file.

**Environment:**
 - Operating system: Windows 10 version 10.0.19041.508
 - Computer: Dell XPS 8920, Intel Core i7-7700, 3.6 GHz, 16 GB ram
 - DOSBox-X:  0.83.5, 4DOS 8.00 reporting MS-DOS 7.10

If you need to see my conf file, let me know and I'll post it.

**Comment by Wengier** [comment]
@zydanielson Thanks for posting the issue here! Unlike vDosPlus, which only had one binary build for each version, DOSBox-X has multiple binary builds for each version (as can be selected from the Windows installer). For example, for the Windows platform there are 32-bit and 64-bit Visual Studio Release builds and 32-bit and 64-bit MinGW builds (either SDL1 or SDL2). They work the same for the most part, but there may be some differences in certain situations depending on your platform so that certain builds work better than the other in these cases. I do not have the C:\BKM\Master.cla file as you mentioned for testing right now, but I did try to use the Clarion Compiler (ccmp.exe) to compile the built-in OPTIONS.CLA source file, and results are like this:

Release 32-bit build: ```NO MESSAGES FOUND IN OPTIONS.CLA``` (compiled correctly)
Release 64-bit build ```517 MESSAGES FOUND IN OPTIONS.CLA``` (compiled errors)
MinGW 32-bit build: ```NO MESSAGES FOUND IN OPTIONS.CLA``` (compiled correctly)
MinGW 32-bit build: ```NO MESSAGES FOUND IN OPTIONS.CLA``` (compiled correctly)

Tested with both DOSBox-X 0.83.4 and 0.83.5, and the result seems to be the same.

I guess this likely applies to your source file too. Could you please try to compile your source files with either the 32-bit release build or the MinGW builds? Thanks!

**Comment by zydanielson** [comment]
I downloaded the portable (zip) versions of MinGW 32-bit and 64-bit. The compiler runs without errors on both portable boxes on my machine. I have not been able to find a portable (zip) version of Windows 32-bit (non-MinGW). Is such a portable zip version available somewhere in the repository? I do not see it listed here:
https://github.com/joncampbell123/dosbox-x/blob/master/INSTALL.md

One additional comment I will make is this. Even though the MinGW builds allows ccmp.exe to compile Clarion 2.1 source files without errors, the compile still runs 29 times slower than when compiling in vDosPlus. I can live with this, but I wonder if you know why this is, and if DOSBox-X will ever match the speed I see in vDosPlus?

**Comment by Wengier** [comment]
@zydanielson Portable versions of Visual Studio builds were not available as zip files for version 0.83.5. If you downloaded the Windows installer, you can select the 32-bit Release version though.

Also, the size of vDosPlus is much smaller than the size of DOSBox-X. Unlike vDosPlus which is only for DOS applications in Windows, DOSBox-X aims to be a cross-platform, complete and accurate DOS emulation package. But it is certain possible to try to improve its speed with the help of perhaps some additional config options later on.

**Comment by Wengier** [comment]
Also, DOSBox-X 0.83.6 is coming later today. It will likely come with separate Visual Studio builds.

**Comment by zydanielson** [comment]
**Where My Project Stands Right Now**
- On my computer (Windows 10, 64-bit), here in my office, I used the 0.83.5 Windows Installer, installing the 64-bit build, then made many modifications to the conf file using the graphical configuration tool.
- At my client's site I did this:
     - Unzipped the portable MinGW 32 build on a Windows 7, 32-bit workstation, then made many modifications to the conf file using the graphical configuration tool.
     - Next I unzipped the same portable build on a Windows 10, 32-bit workstation and copied the conf file over from the Windows 7 machine.
     - These machines are part of a Windows Server 2016 network. The Clarion 2.1 source, data, and processor files are on the server in a shared directory (\\\BKM-2016\BKMsanbx). Both computers mount this resource in DOSBox-X as Q: drive. ***The Clarion system runs great*** on both machines with multi-user issues already handled in the Clarion code. There are some problems that are related to my Clarion code that I am still working through, but this has nothing to do with DOSBox-X. I'm confident I'll get everything to work as it should using the Clarion 2.1 runtime (crun.exe). ***This is a very successful first step!***
- Now I would like to upgrade to 0.83.6. I've seen a lot of information in the Wiki on _installing_ DOSBox-X, but haven't found much on _upgrading_ it. (Maybe I just haven't looked in the right place.) So I have a few questions.

**Questions About Upgrading**
1. If I upgrade my computer using the Windows installer, do I first have to uninstall the previous installation before running the Windows 0.83.6 installer? Or can I run it and install right over the previous installation?
1. When does the conf file get upgraded, preserving the changes I've already made to it with the previous configuration tool? Does it happen the first time I use the configuration tool in the new installation? (i.e., is it the configuration tool that upgrades the conf file it finds even if it was created in a previous version?)
1. Next is the decision on how to upgrade my client's computers. If I download the new portable MinGW 32-bit build and unzip the correct DOXBox-X.exe into the directory where the present conf file is stored, I assume this will run the new version of DOXBox-X, right? But how will the conf files on those machines get upgraded, preserving the changes I've already made? Just by running the configuration tool in the new version?

Once I successfully upgrade DOSBox-X on my machine and my client's machines, I'll resume testing.

**Comment by Wengier** [comment]
The Windows installer for DOSBox-X is very powerful. It is the recommended way to install DOSBox-X or upgrade to a new version of DOSBox-X, and there is definitely no need to uninstall the previous version before running the new installer. The Windows installer offers an option to automatically upgrade the config file (dosbox-x.conf) to the new version format while keeping all the user-customized settings already made. If you are not 100% sure, you can backup the config file first and then run the Windows installer and let it automatically upgrade the config file.

As for your client's computer, you can do the similar thing too. Make sure the config file is named dosbox-x.conf in the DOSBox-X directory, and then run Windows installer and install it to that directory. Again select the option to upgrade the config file to the new version format while keeping existing user settings, and it should work.

The built-in configuration tool on the other hand has no ability to upgrade existing config files. Only the Windows installer can do this at this time.

And you are right that the information about upgrading was not mentioned in the Wiki or the INSTALL page. I have already added it to the INSTALL page. Thanks for mentioning this!

**Comment by zydanielson** [comment]
I have used the windows installer, selecting the 32-bit build as the default on my machine. I'm happy to report that I am now successfully running DOSBox-X 0.83.6. The best news of all is that the Clarion 2.1 Compiler (ccmp.exe) runs without creating false error messages on my Windows 10, 64-bit computer. The installer also upgraded my conf file, as designed, keeping my previous settings intact. Thank you for this excellent work.

It is important for a Clarion developer to remember that it doesn't matter whether the computer where DOSBox-X is running is a 64-bit machine. What matters is that the Compiler (ccmp.exe) will not run in a 64-bit build of DOSBox-X. Therefore, anyone running the Clarion 2.1 Professional Developer should install the Win-32 build.

Before we close this issue, I will use the Windows installer to upgrade the two computers on my client's system to the Win-32 build. Both of these machines were using the MinGW-32 portables. I will create a DOSBox-X directory on the C: drive of each machine and copy the conf file from the portable directory into it. Then I'll run the installer choosing that directory to install into. Even though that directory will contain nothing but the conf file, I believe the install should work, finding and upgrading the conf file as well. When that is successfully tested, we'll be able to close this issue. That should happen by late Friday afternoon.

**Comment by zydanielson** [comment]
Both installations worked great at my client's site. I think we can close this issue now. The reason I didn't click the **Close with comment** button is because I would like to know if a closed issue can be reopened in the future if something else comes up with it? Perhaps just posting a comment to a closed issue will reopen it? Let me know.

Thanks for all your help on this, @Wengier.

**Comment by Wengier** [comment]
@zydanielson Thanks for reporting the good results! Sure, you can close and then reopen an issue at any time when you feel it is necessary. You can also post a comment to a closed issue without reopening it. So it is pretty flexible.

**Comment by zydanielson** [comment]
Got it. Thanks.   🥇

**Comment by zydanielson** [comment]
### Another problem in migrating from vDosPlus to DOSBos-X.
This problem has nothing to do with the compiler, but a single Clarion 2.1 program I've written behaves differently in vDosPlus than it does in DOXBox-X. Unfortunately, the desired (and necessary) behavior happens in all virtual boxes tested, but does _not_ happen in DOXBox-X.

### The Clarion 2.1 Program
The program shells out to DOS to run a batch file called Message.bat. Here is the content of Message.bat:

> `@ECHO OFF`
> `CLS`
> `IF NOT EXIST MESSAGE.TXT GOTO ENDLINE`
> `TYPE MESSAGE.TXT`
> `PAUSE`
> `:ENDLINE`

I wrote this a long time ago to pass short messages to the Brecon operators when the Master Menu loads, but before the menu displays. All I had to do to activate the message was create a MESSAGE.TXT file in the root directory. The key command was the `PAUSE` in the batch file. The user sees the message and presses a key to continue. The shell returns control to the Clarion program, which then displays the Master Menu.

When I no longer wanted the user to see the message, I simply deleted MESSAGE.TXT and the batch file bypassed the `PAUSE` command by jumping to `:ENDLINE` immediately returning control to the Clarion program, which displayed the Master Menu.

### When running correctly in vDosPlus, VMware, and Oracle VirtualBox
The program works as designed. The user sees:
![ShellOutToBatWithPause](https://user-images.githubusercontent.com/11687580/95400877-03322a00-08c9-11eb-84e8-924f1683d3bb.jpg)
The user presses a key and the Master Menu appears.

### When running _incorrectly_ in DOSBox-X
The screen blinks quickly and the Master Menu immediately displays. Apparently when the program shells out to run the batch file, DOSBox-X isn't waiting for the batch file to complete, but rather launches it and immediately returns. The user never gets a chance to read MESSAGE.TXT and press a key.

If there is a setting somewhere to fix this problem, I haven't found it.

**Comment by Wengier** [comment]
Could you upload the said Clarion 2.1 Program so that we can have a test for this? Thanks!

**Comment by zydanielson** [comment]
I'll zip several files together in a compressed folder so that you'll have all you need to reproduce the problem. I just need to know if I upload files by dragging them in just like dragging and dropping an image? Or is there some other place or way to upload files that aren't images?

**Comment by Wengier** [comment]
@zydanielson I think you can drag and drop zip files here just like dragging and dropping an image. Dragging and dropping executable files on the other hand may not work in the reply form.

**Comment by zydanielson** [comment]
@Wengier  Here is the zipped folder containing all you need for testing.

[CLA2.1zyDanielson.zip](https://github.com/joncampbell123/dosbox-x/files/5346159/CLA2.1zyDanielson.zip)

**Comment by Wengier** [comment]
@zydanielson I have tested the zip package you uploaded and extracted to a directory, but in both vDosPlus and DOSBox-X the following error box is shown when run:

```
                       Run Time Halt
CRUN Version 2.1 Release 2105 10/08/120  8:06 PM
C:\CLATEST\SETKCD.BIN
ERROR CODE:19 MODULE NOT FOUND
DOS ERROR :2 FILE NOT FOUND
PROGRAM   :C:\CLATEST\MASTER.PRO
MODULE    :No Program Loaded
HALTED AT :
CALL FROM :
                 Press any key to continue
```

And there is apparently no file named ```SETKCD.BIN``` in the zip package. Can you please check that?

**Comment by zydanielson** [comment]
@Wengier, sorry about that. Two resources were loading from a path I had set in the environment. I have included them both now in this new zip of ClaTest.

[CLA2.1zyDanielson2.zip](https://github.com/joncampbell123/dosbox-x/files/5352022/CLA2.1zyDanielson2.zip)

I've also updated the ReadMe.txt file inside ClaTest in case you want to give that another read.

**Comment by Wengier** [comment]
@zydanielson I have tried the new zip package you uploaded, the running "crun master" in DOSBox-X 0.83.6 will get the following screen:

![image](https://user-images.githubusercontent.com/8216923/95540967-2095ef80-09c0-11eb-8573-b49ecafdbb3e.png)

And if I press any key, the following screen will show up:

![image](https://user-images.githubusercontent.com/8216923/95541040-47ecbc80-09c0-11eb-8848-6075da15944d.png)

Seems that the result is basically the same as in vDosPlus. I tried with the default setting (dosbox-x -defaultconf). Perhaps you can try this too, and check if anything abnormal?

**Comment by zydanielson** [comment]
Okay, that worked for me also. However, I need some of the conf modifications I made in order to get the full Brecon system to run smoothly. The only thing I know to do is use WinMerge on the conf file I modified beside the default conf file to see which settings are different. That could cost me quite a bit of time, turning the differing options off and on one at a time until the problem stops. It seems like a lot of work. But I really do need to know which setting(s) caused this.

Do you have a better idea of which setting might be a problem so that you can point me in a good direction and help me avoid pure trial and error? If you're still willing to help, I could send you all the differences WinMerge finds in the two conf files and perhaps you could suggest which setting is or which settings are the most likely culprit(s)?

**Comment by Wengier** [comment]
@zydanielson You can also upload your conf file and let me see which setting in your conf file may have caused the said problem.

**Comment by zydanielson** [comment]
I found it. I had set **shell = 4DOS.COM** and when I reset it to the default **shell =** it started working as expected.

I don't really understand why the 4DOS.COM shell won't wait for a batch file to complete before returning, but this is what caused it.

**Comment by Wengier** [comment]
Thanks for reporting this! I can reproduce the issue here with ```shell=4dos.com``` and I have hopefully fixed it in the latest code. You can download the updated Windows binary here:

Download link: [dosbox-x-update.zip](https://github.com/Wengier/dosbox-x-wiki/blob/master/dosbox-x-update.zip?raw=true)

**Comment by zydanielson** [comment]
@Wengier: confirmed. The code change works correctly in my machines with `shell=4dos.com`

Excellent support. Thank you.

**Comment by zydanielson** [comment]
@joncampbell123, @Wengier, _et. al._

I just wanted to report that the Brecon Knitting Mill's system has been thoroughly tested and is now running live, completely in DOSBox-X. This is quite a feather in the cap for the DOSBox-X project. The Brecon system is a huge, fully integrated industrial package written in Clarion 2.1 that has been in steady use for over 30 years. Now Brecon can continue to use it into the future, regardless that DOS applications are no longer supported by Microsoft.

Brecon will continue to keep their VMware for the time being, but only as a backup in the unlikely event that something in DOSBox-X should fail. But the VMware will be turned OFF, which should speed up the performance of the server. After a month without problems, Allen will have his Windows Administrators uninstall the VMware entirely. Both Allen and I are very happy with this development and thank you for making it possible.

There is only one very small "cosmetic" problem we've discovered with DOSBox-X. But I will open a new issue to report that and it isn't a high priority as functionality is not affected.

You have made a huge contribution to my client and I will remain in perpetual gratitude for as long as I continue to program and support Brecon. I'm almost 68 now. I've been programming and supporting Brecon for roughly 46 years.

Sincerely yours,
Zy Danielson

**Comment by Wengier** [comment]
@zydanielson Glad to hear that the system is fully working with DOSBox-X! It is certainly great that you can now run Brecon in DOSBox-X just like how it worked in DOS, regardless that DOS applications are no longer supported by Microsoft as you said. I hope to look into the "cosmetic" problem you mentioned as well.
